[ZEPPELIN-2949] Allow custom Spark UI URL

necosta opened the pull request:
### What is this PR for?
Allow an override of default Spark logic in building the URL path and redirect the Zeppelin user directly to a configured reverse proxy path

### What type of PR is it?
[Improvement]

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-2949

### How should this be tested?
* Set "zeppelin.spark.uiWebUrl" to http://localhost/spark-reverse-proxy-path
* Run Spark code
* Open Spark UI
* Expected to be redirected to http://localhost/spark-reverse-proxy-path

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? N
* Is there breaking changes for older versions? N
* Does this needs documentation? N

necosta commented:
Thanks @felixcheung. Test looks unrelated but I'll double check.
Added one commit for documentation

Not sure how much we want to document on how to work with Spark behind reverse proxy but I can add that by setting properties:
* spark.ui.enabled=true
* spark.ui.reverseProxy=false (disables '/proxy/applicationId' web url path)
* spark.ui.proxyBase=/reverseProxyUniqueKey
* zeppelin.spark.uiWebUrl=http://{hostName}/reverseProxyUniqueKey

https://issues.apache.org/jira/browse/SPARK-20044 is preventing us from using a path prefix and it is assumed we are routing from spark port to reverseProxyUniqueKey

necosta commented:
Not sure what we can do to have the tests passing:
https://travis-ci.org/nokia/zeppelin/jobs/279484561
https://travis-ci.org/nokia/zeppelin/jobs/279484562
Any suggestions?

felixcheung commented on docs/interpreter/spark.md:
```diff
@@ -145,6 +145,11 @@ You can also set other Spark properties which are not listed in the table. For a
+  <tr>
+  <td>zeppelin.spark.uiWebUrl</td>
+    <td></td>
+    <td>Override Spark UI default URL</td>
```
could you add a bit more on what are the condition to set this, example value etc..?

necosta commented:
Thanks @felixcheung and apologies for the delay on this.
Rebased on top of latest master and updated documentation.
Can you have another look?
Thanks

felixcheung commented on docs/interpreter/spark.md:
```diff
@@ -145,6 +145,11 @@ You can also set other Spark properties which are not listed in the table. For a
+  <tr>
+  <td>zeppelin.spark.uiWebUrl</td>
+    <td></td>
+    <td>Overrides Spark UI default URL (ex: http://{hostName}/{uniquePath)</td>
```
should this be `{uniquePath)` -> `{uniquePath})`

felixcheung replied:
I think we should clarify that the value should be a *full* URL

necosta replied:
Fixed on latest commit

felixcheung commented on spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java:
```diff
@@ -936,6 +936,11 @@ public String getSparkUIUrl() {
     }
 
+    String sparkUrlProp = property.getProperty("zeppelin.spark.uiWebUrl", "");
+    if (!sparkUrlProp.isEmpty()) {
```
let's check for isBlank https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isBlank-java.lang.CharSequence-

necosta replied:
I agree. Fixed on latest commit

felixcheung commented:
merging if no more comment
